06-TimeManagement/omod6_kern_time.c

```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: AFL-3.0
+/*
+ * Copyright (C) 2018
+ * Author: Oleg Khokhlov <oleg.khokhlov.ua@gmail.com>
+ *
+ * SysFS kernel module for test kernel timing functions
+ *
+ * doc links:
+ *
+ */
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/device.h>
+#include <linux/err.h>
+#include <linux/uaccess.h>
+#include <linux/string.h>
+
+#include <linux/jiffies.h>
+#include <linux/ktime.h>
+#include <linux/timer.h>
+#include <linux/rtc.h>
+#include <linux/delay.h>
+
+//Internal Kernel API (Time Management):
+
+//1. Implement kernel module with API in sysfs or procfs, which is able to:
+// - return time (in seconds) passed since previous read;
+// - return absolute time of previous reading;
+// - periodically print some message to log, make period adjustable;
+
+// --------------- MODULE PARAMETERS DESCRIPTION -------------------
+//int iParam;
+//module_param(iParam, int, 0);
+//MODULE_PARM_DESC(iParam, "iParam: ReturnCode");
+
+//char *sParam = "Default";
+//module_param(sParam, charp, 0);
+//MODULE_PARM_DESC(sParam, "sParam: string parameter");
+
+//------------ GLOBAL MODULE DATA ---------------------
+static struct class *attr_class;
+
+#define SysFS_DIR_NAME   "omod6"
+#define SysFS_ATTR_NAME  "ktime"
+
+uint32_t ModCallCount;
+uint32_t LastCallTimeSec;
+uint32_t LastCall_dT;
+uint32_t LastCallTicks;
+
+uint32_t TimerInterval;
+uint32_t TimerCounter;
+uint32_t TimerStartTick;
+
+char LogMessage[PAGE_SIZE];
+
+uint32_t GetTimeStamp(void)
+{
+	return (uint32_t) (ktime_to_ms(ktime_get_real()) / 1000);
+}
+
+uint32_t GetTickCount(void)
+{
+	return (uint32_t) (get_jiffies_64() * 1000 / HZ);
+}
+
+
+//--------------- SysFS read/write functions ----------------------
+
+static ssize_t ktime_show(struct class *cl,
+			struct class_attribute *attr,
+			char *buf)
+{
+	int L;
+
+	uint32_t TS, dT, T;
+
+	T = GetTickCount();
+	TS = GetTimeStamp();
+	dT = TS - LastCallTimeSec;
+
+	if (dT == 0)
+		dT = LastCall_dT;
+	else
+		LastCall_dT = dT;
+
+	LastCallTicks = T;
+
+	ModCallCount++;
+
+	sprintf(buf, "ktime (%u ticks, %u HZ):\n", (uint32_t)get_jiffies_64(), HZ);
+	sprintf(strchr(buf, 0), "%u module call\n", ModCallCount);
+	sprintf(strchr(buf, 0), "%u sec from last call\n", dT);
+	sprintf(strchr(buf, 0), "%u sec curr abs time from Epoch (year:%d)\n", TS, 1970+TS/(3600*24*365));
+	sprintf(strchr(buf, 0), "%u sec prev call abs time from Epoch\n", LastCallTimeSec);
```

Comment:
> 1542096137 sec curr abs time from Epoch (year:2018)

IMO it's better to format absolute time stamp as human readable string aka YY.MM.DD hh:mm:ss

Reply:
Good idea. I had been thinking about it when I was written the code but I decided that it's extraneous. Maybe you can recomend proper kernel function to convert epoch timestamp to struct time ?

Reply:
I see you've managed to format the epoch time. Good job!